### ISSUE TITLE
Person Counting

### ISSUE THREAD

**Issue by amir2802** [AlexeyAB/darknet#5090]
Dear AlexeyAB,

I really need your help. Where and how should i modify coding to get output video in counting person instead of detection person. I am running in Google Colab. I am trying to get the ouput similar like photo below. Really appreciate your help.

![IMG_20200323_110118](https://user-images.githubusercontent.com/62529323/77276452-ae94dc00-6cf5-11ea-9be5-6d7d68ace94c.JPG)

**Comment by flowzen1337** [comment]
Which code do you use, the darknet.py / darknet_video.py from this repo? Then you got the "detection" array in that python, where all the detections for your frame are in. You can iterate over this array (done in the cvDrawBoxes function already) and add +1 to a counter whenever a object "person" is found. Then just cv2.putText... this str(counter) on your frame at the end (but inside!) the while loop and et voila, you got the person counter :-)

**Comment by amir2802** [comment]
Yes, i am using  the darknet.py / darknet_video.py from your repo.
I have found the function that you mention (cvDrawBoxes). But due to my poor knowledge in coding and deep learning, honestly i am still stuck in implementing the counting's command into the coding. Sorry for my lack of understanding. Hopefully you can guide me to implement the coding in simple way. Really sorry again. 🙏

![Screenshot_20200325_101158](https://user-images.githubusercontent.com/62529323/77495021-61924080-6e82-11ea-8607-b477cfcaa1df.jpg)

**Comment by flowzen1337** [comment]
Ok here is the modified cvDrawBoxes function (added comments for parts i've added):

`def cvDrawBoxes(detections, img):

    # initilalize the person counter variable
    pc=0

    for detection in detections:
        x, y, w, h = detection[2][0],\
            detection[2][1],\
            detection[2][2],\
            detection[2][3]
        xmin, ymin, xmax, ymax = convertBack(
            float(x), float(y), float(w), float(h))
        pt1 = (xmin, ymin)
        pt2 = (xmax, ymax)

        # add +1 to our person counter if we found a object 'person'
        if str(detection[0].decode()) == 'person':
                pc+=1

        cv2.rectangle(img, pt1, pt2, (0, 255, 0), 1)
        cv2.putText(img,
                    detection[0].decode() +
                    " [" + str(round(detection[1] * 100, 2)) + "]",
                    (pt1[0], pt1[1] - 5), cv2.FONT_HERSHEY_SIMPLEX, 0.5,
                    [0, 255, 0], 2)

    # return image and person counter
    return img, pc`


Outside of the function, in the main code line +106, consume our function return image and pc:
`....
....
# added person counter to read the pc return of  the cvDrawBoxes function
image, pc = cvDrawBoxes(detections, frame_resized)

# added cv2.putText to draw the counter on our image
cv2.putText(img, "Persons: "+str(pc), (25, 25), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0,125,255), 2)
....
....`

It's untested but give it a try, i guess it should work :-)

**Comment by amir2802** [comment]
hi alexey, sorry for late reply due to MCO COVID19..
actually, i have some questions..
regarding these 2 codings below that u helped me previously,
should i *edit at the existing* coding OR i *need to add new* to the
existing coding in darknet_video.py?

added person counter to read the pc return of the cvDrawBoxes function

image, pc = cvDrawBoxes(detections, frame_resized)

*existing coding :*
while True:
        prev_time = time.time()
        ret, frame_read = cap.read()
        frame_rgb = cv2.cvtColor(frame_read, cv2.COLOR_BGR2RGB)
        frame_resized = cv2.resize(frame_rgb,
                                   (darknet.network_width(netMain),
                                    darknet.network_height(netMain)),
                                   interpolation=cv2.INTER_LINEAR)

        darknet.copy_image_from_bytes(darknet_image,frame_resized.tobytes())

        detections = darknet.detect_image(netMain, metaMain, darknet_image,
 thresh=0.25)
        image = cvDrawBoxes(detections, frame_resized)
        image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
        print(1/(time.time()-prev_time))
        cv2.imshow('Demo', image)
        cv2.waitKey(3)
    cap.release()
    out.release()

added cv2.putText to draw the counter on our image

cv2.putText(img, "Persons: "+str(pc), (25, 25), cv2.FONT_HERSHEY_SIMPLEX,
0.5, (0,125,255), 2)

*existing coding : *
def cvDrawBoxes(detections, img):
    for detection in detections:
        x, y, w, h = detection[2][0],\
            detection[2][1],\
            detection[2][2],\
            detection[2][3]
        xmin, ymin, xmax, ymax = convertBack(
            float(x), float(y), float(w), float(h))
        pt1 = (xmin, ymin)
        pt2 = (xmax, ymax)
        cv2.rectangle(img, pt1, pt2, (0, 255, 0), 1)
        cv2.putText(img,
                    detection[0].decode() +
                    " [" + str(round(detection[1] * 100, 2)) + "]",
                    (pt1[0], pt1[1] - 5), cv2.FONT_HERSHEY_SIMPLEX, 0.5,
                    [0, 255, 0], 2)
    return img

really appreciate and thankful for your kind help.

**Comment by amir2802** [comment]
*Hi AlexeyAB,*

*i already run below modified command for darknet_video.py but
unfortunately the result still only detection of person, no counting of
person appear.. yellow highlighted are command that i added.. sorry for my
lack of understanding and knowledge..*

from ctypes import *
import math
import random
import os
import cv2
import numpy as np
import time
import darknet

def convertBack(x, y, w, h):
    xmin = int(round(x - (w / 2)))
    xmax = int(round(x + (w / 2)))
    ymin = int(round(y - (h / 2)))
    ymax = int(round(y + (h / 2)))
    return xmin, ymin, xmax, ymax


def cvDrawBoxes(detections, img):

pc=0

    for detection in detections:
        x, y, w, h = detection[2][0],\
            detection[2][1],\
            detection[2][2],\
            detection[2][3]
        xmin, ymin, xmax, ymax = convertBack(
            float(x), float(y), float(w), float(h))
        pt1 = (xmin, ymin)
        pt2 = (xmax, ymax)

    if str(detection[0].decode()) == 'person':
            pc+=1

        cv2.rectangle(img, pt1, pt2, (0, 255, 0), 1)
        cv2.putText(img,
                    detection[0].decode() +
                    " [" + str(round(detection[1] * 100, 2)) + "]",
                    (pt1[0], pt1[1] - 5), cv2.FONT_HERSHEY_SIMPLEX, 0.5,
                    [0, 255, 0], 2)
    return img, pc


netMain = None
metaMain = None
altNames = None


def YOLO():

    global metaMain, netMain, altNames
    configPath = "./cfg/yolov4.cfg"
    weightPath = "./yolov4.weights"
    metaPath = "./cfg/coco.data"
    if not os.path.exists(configPath):
        raise ValueError("Invalid config path `" +
                         os.path.abspath(configPath)+"`")
    if not os.path.exists(weightPath):
        raise ValueError("Invalid weight path `" +
                         os.path.abspath(weightPath)+"`")
    if not os.path.exists(metaPath):
        raise ValueError("Invalid data file path `" +
                         os.path.abspath(metaPath)+"`")
    if netMain is None:
        netMain = darknet.load_net_custom(configPath.encode(
            "ascii"), weightPath.encode("ascii"), 0, 1)  # batch size = 1
    if metaMain is None:
        metaMain = darknet.load_meta(metaPath.encode("ascii"))
    if altNames is None:
        try:
            with open(metaPath) as metaFH:
                metaContents = metaFH.read()
                import re
                match = re.search("names *= *(.*)$", metaContents,
                                  re.IGNORECASE | re.MULTILINE)
                if match:
                    result = match.group(1)
                else:
                    result = None
                try:
                    if os.path.exists(result):
                        with open(result) as namesFH:
                            namesList = namesFH.read().strip().split("\n")
                            altNames = [x.strip() for x in namesList]
                except TypeError:
                    pass
        except Exception:
            pass
    #cap = cv2.VideoCapture(0)
    cap = cv2.VideoCapture("test.mp4")
    cap.set(3, 1280)
    cap.set(4, 720)
    out = cv2.VideoWriter(
        "output.avi", cv2.VideoWriter_fourcc(*"MJPG"), 10.0,
        (darknet.network_width(netMain), darknet.network_height(netMain)))
    print("Starting the YOLO loop...")

    # Create an image we reuse for each detect
    darknet_image = darknet.make_image(darknet.network_width(netMain),
                                    darknet.network_height(netMain),3)
    while True:
        prev_time = time.time()
        ret, frame_read = cap.read()
        frame_rgb = cv2.cvtColor(frame_read, cv2.COLOR_BGR2RGB)
        frame_resized = cv2.resize(frame_rgb,
                                   (darknet.network_width(netMain),
                                    darknet.network_height(netMain)),
                                   interpolation=cv2.INTER_LINEAR)

        darknet.copy_image_from_bytes(darknet_image,frame_resized.tobytes())

        detections = darknet.detect_image(netMain, metaMain, darknet_image,
thresh=0.25)
        image = cvDrawBoxes(detections, frame_resized)
        image, pc = cvDrawBoxes(detections, frame_resized)
        image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
        print(1/(time.time()-prev_time))
        cv2.imshow('Demo', image)
        cv2.waitKey(3)
        cv2.putText(img, "Persons: "+str(pc), (25, 25),
cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0,125,255), 2)
    cap.release()
    out.release()

if __name__ == "__main__":
    YOLO()
